File: problems/linked-list/0141-linked-list-cycle.js
/**
 * #141 Linked List Cycle
 * Pattern: Floyd's fast and slow pointers.
 * If a cycle exists, the pointers must meet inside it.
 * Time: O(n) | Space: O(1)
 *
 * @param {ListNode|null} head
 * @return {boolean}
 */
function hasCycle(head) {
  let slow = head
  let fast = head

  while (fast !== null && fast.next !== null) {
    slow = slow.next
    fast = fast.next.next
    if (slow === fast) return true
  }

  return false
}

module.exports = hasCycle

File: problems/linked-list/0141-linked-list-cycle.py
"""#141 Linked List Cycle
Floyd's pointers meet if the fast pointer laps the slow pointer in a cycle.
Time: O(n) | Space: O(1)
"""


class Solution:
    def hasCycle(self, head):
        slow = fast = head
        while fast is not None and fast.next is not None:
            slow = slow.next
            fast = fast.next.next
            if slow is fast:
                return True
        return False
